feat(api): Add api exception mapping

tatarco opened the pull request:
### What changed? Why was the change needed?
We are having issues caused by a very basic error handling in our nest.js app, causing 500 error to not be logged in a good way, the solution is described in the loom here 
https://www.loom.com/share/e6d13666ce94441c9316e9f13bd94472?sid=17c07829-447c-40ac-a9bd-9e01279729ad

<!-- Also include any relevant links, such as Linear tickets, Slack discussions, or design documents. -->

### Screenshots
<!-- If the changes are visual, include screenshots or screencasts. -->

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Related enterprise PR
<!-- A link to a dependent pull request  -->

### Special notes for your reviewer
<!-- Specific instructions or considerations you want to highlight for the reviewer. -->

</details>

tatarco committed:
feat(api): add exception mapper + sentry updates

tatarco commented on apps/api/src/app/workflows-v2/workflow.controller.e2e.ts:
```diff
@@ -24,6 +24,7 @@ const TEST_TAGS = ['test'];
 let session: UserSession;
 
 const SCHEMA_WITH_TEXT: JsonSchema = {
+  $schema: 'http://json-schema.org/draft-07/schema#',
```
`  
`

SokratisVidros commented on apps/api/src/bootstrap.ts:
```diff
@@ -121,7 +122,8 @@ export async function bootstrap(expressApp?): Promise<INestApplication> {
   Logger.log('BOOTSTRAPPED SUCCESSFULLY');
-
+  app.useGlobalFilters(new AllExceptionsFilter(app.get(PinoLogger)));
+  Logger.log('Filter Applied');
```
Can we move this comment into the constructor of the exception filter? I guess it's instantiated once on bootstrapping. That way we will have everything contained.

tatarco replied:
removed

SokratisVidros commented on apps/api/src/bootstrap.ts:
```diff
@@ -121,7 +122,8 @@ export async function bootstrap(expressApp?): Promise<INestApplication> {
 
   Logger.log('BOOTSTRAPPED SUCCESSFULLY');
-
+  app.useGlobalFilters(new AllExceptionsFilter(app.get(PinoLogger)));
```
@tatarco The PR looks great! I expected to add this but remove the current handling of exceptions and erroneous response bodies. Where are we generating response 5xx and 4xx responses today? 

Do we need to shortcircuit that or remove the code alltogether?

tatarco replied:
Richard and I did some research, this does not damage the existing functionality, I will attempt to use the sentry ID if possible, currently we are using the default simplistic node.js error mapping which is just mapping 500 for unexpected exceptions and map other apiError's with their constructor values.

SokratisVidros commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,87 @@
+
+  private buildResponseBody(status: number, request: Request, message: string | object | Object, exception: unknown) {
+    let responseBody = this.buildBaseResponseBody(status, request, message);
+    if (status === HttpStatus.INTERNAL_SERVER_ERROR) {
```
The response body should be constructed for all 5xx errors. 

```suggestion
    if (status >= HttpStatus.INTERNAL_SERVER_ERROR) {
```

rifont commented on apps/api/src/bootstrap.ts:
```diff
@@ -121,7 +122,8 @@ export async function bootstrap(expressApp?): Promise<INestApplication> {
   await setupSwagger(app);
 
   Logger.log('BOOTSTRAPPED SUCCESSFULLY');
-
+  app.useGlobalFilters(new AllExceptionsFilter(app.get(PinoLogger)));
+  Logger.log('Filter Applied');
   if (expressApp) {
```
The successful bootstrap log should be the last one we see to indicate the app has fully started up - if filter application fails, we could be mislead that the application started when it failed right at the end.

✔️ chore: Move these 2 newly added lines above the bootstrap message

rifont commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,87 @@
+    response.status(status).json(responseBOdy);
+  }
+
+  private buildResponseBody(status: number, request: Request, message: string | object | Object, exception: unknown) {
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common process. Try to leave a link to the process description so that the reader knows how to resolve the chore.  -->
☑ **chore:** Please add a response type to all methods‏

rifont commented on apps/api/src/bootstrap.ts:
```diff
@@ -121,7 +122,8 @@ export async function bootstrap(expressApp?): Promise<INestApplication> {
   await setupSwagger(app);
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
❓ **question:** Please see [line 93](https://github.com/novuhq/novu/pull/6601/files#diff-11933634457bf1519b686a8747a356d1c21a23aa19c22984b3da6c875919c6c3R93) above, we already have Sentry instrumentation to handle exceptions. We should make sure that we're handling exceptions consistently with a single entry point.

Related: it seems that the latest version of `@sentry/node` removed the `Handlers` export, [see here for more details on migrating](https://github.com/getsentry/sentry-javascript/issues/12007). We should explore this upgrade, it's possible that the newer versions of Sentry have built-in APIs for exception mapping.‏

tatarco replied:
for context: we have synced and tested the change with sentry, the decision is to attempt to match the sentry error id with the uuid if exist (for cloud) besides that this functionality works great and does not damage the sentry functionality

rifont commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,87 @@
+
+  // Construct the UUID using the random hex values
+  return `${randomHex()}${randomHex()}-${randomHex()}-${randomHex()}-${randomHex()}-${randomHex()}${randomHex()}${randomHex()}`;
+}
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution.  -->
❓ **question:** `node:crypto` has a [`randomUUID`](https://developer.mozilla.org/en-US/docs/Web/API/Crypto/randomUUID)‏ function as of Node v19, is there a specific reason for rolling our own UUID generator? If there is, please add a code comment explaining why. Otherwise we can replace with `randomUUID`.

tatarco replied:
there was some git secutiry issue with it, i'll change

rifont commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,87 @@
+import { ArgumentsHost, Catch, ExceptionFilter, HttpException, HttpStatus } from '@nestjs/common';
+import { Response } from 'express';
+import { randomBytes } from 'crypto';
```
<!-- Chores are simple tasks that must be done before the subject can be “officially” accepted. Usually, these comments reference some common process. Try to leave a link to the process description so that the reader knows how to resolve the chore.  -->
☑ **chore:** Please use the `node:` import directive for built-in modules to avoid the risk of colliding with modules in `node_modules`. It also helps to draw attention to the fact that a built-in module is being used. You can read more about this directive [here](https://2ality.com/2021/12/node-protocol-imports.html).‏
```suggestion
import { randomBytes } from 'node:crypto';
```

rifont reviewed (commented):
The direction is a valid one, let's verify firstly if Sentry can provide exception mappers that provide the necessary functionality and proceed from there.

rifont commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,87 @@
+      const uuid = generateUUID();
+      this.pinoLogger.error(`[${uuid}] Service thrown an unexpected exception: `, formatError(exception));
+
+      return { ...responseBody, errorId: uuid };
```
As discussed:

We verified that Sentry reports all errors when `SENTRY_DSN` environment variable is supplied.

See if we can use the Sentry ID if it exists (see reference error in [Sentry here](https://novu-r9.sentry.io/issues/5945392524/events/5e1de89d6eeb4250b092781b8beadb2e/)) [Cloud Edition], otherwise generate a UUID to support Community Edition.

BiswaViraj committed:
feat(js): Com 229 update the in app preview component in the web app to (#6600)

Co-authored-by: Sokratis Vidros <SokratisVidros@users.noreply.github.com>

LetItRock committed:
feat(dashboard): side navigation (#6608)

SokratisVidros committed:
fix(api,web): Align segment traits payload (#6628)

SokratisVidros committed:
fix(api): Update .source and quota-throttler tests (#6627)

BiswaViraj committed:
feat(js, react, shared): user agents (#6626)

rifont committed:
chore(root): Prevent multi-level imports for `@novu/*` packages (#6596)

LetItRock committed:
chore(dashboard): free trial card (#6612)

djabarovgeorge commented on apps/api/src/app/testing/product-feature.e2e.ts:
```diff
@@ -27,6 +27,6 @@ describe('Product feature Test @skip-in-ee', async () => {
   it('should return a 402 response when required api service level does not exists on organization for feature', async () => {
     const { body } = await session.testAgent.get(path).set('authorization', `ApiKey ${session.apiKey}`).expect(402);
-    expect(body).to.deep.equal({ statusCode: 402, message: 'Payment Required' });
+    expect(body.statusCode).to.equal(402);
```
Can we bring back the `message` expect just in case it will break?

djabarovgeorge commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,93 @@
+    const request = ctx.getRequest<Request>();
+
+    const { status, message } = this.getResponseMetadata(exception);
+    const responseBOdy = this.buildResponseBody(status, request, message, exception);
```
```suggestion
    const responseBody = this.buildResponseBody(status, request, message, exception);
```

djabarovgeorge commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,93 @@
+    let responseBody = this.buildBaseResponseBody(status, request, message);
+    if (status === HttpStatus.INTERNAL_SERVER_ERROR) {
+      const uuid = getUuid(exception);
+      this.pinoLogger.error(`[${uuid}] Service thrown an unexpected exception: `, formatError(exception));
```
Mentioned below that maybe it will be beneficial to change the error type and params order.

tatarco replied:
i don't understand what you want here

djabarovgeorge commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,93 @@
+
+function formatError(error: unknown): string {
+  if (error instanceof Error) {
+    return `
+      Error: ${error.message}
+      Stack: ${error.stack || 'No stack trace available'}
+    `;
+  } else {
+    return `
+      Unknown error: ${JSON.stringify(error, null, 2)}
+    `;
```
If we create an object and pass it as the first param in the logger error function `Logger.error(errorObject, message, context)` it will be parsed as separate params in New Relic see picture below. what do you think?

![image](https://github.com/user-attachments/assets/c5981427-f7e8-48fd-aee3-b667537604a4)


```suggestion
    return {
      Error: ${error.message}
      Stack: ${error.stack || 'No stack trace available'}
    };
  } else {
    return {
    Error:
      Unknown error: ${JSON.stringify(error, null, 2)}
    };
```

tatarco replied:
i'm trying to make sure it's working locally good, if so i'll commit it

djabarovgeorge commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,93 @@
+    `;
+  } else {
+    return `
+      Unknown error: ${JSON.stringify(error, null, 2)}
```
Should we guard ourselves here from 3rd party potential errors with circular reference?

tatarco replied:
i'll add, could be usefull

djabarovgeorge commented on apps/api/src/exception-filter.ts:
```diff
@@ -0,0 +1,93 @@
+export interface ErrorDto {
+  statusCode: number;
+  timestamp: string;
+  errorId?: string;
```
Let's add a comment explaining what the `errorId` is, so future maintainers can understand its purpose and why it's optional.

tatarco replied:
done

djabarovgeorge commented on apps/ws/src/app.module.ts:
```diff
@@ -1,22 +1,20 @@
   LoggerModule,
   ProfilingModule,
-  TracingModule,
-} from '@novu/application-generic';
+  TracingModule
+} from "@novu/application-generic";
 
-import { AppController } from './app.controller';
-import { AppService } from './app.service';
-import { SharedModule } from './shared/shared.module';
-import { HealthModule } from './health/health.module';
-import { SocketModule } from './socket/socket.module';
+import { AppController } from "./app.controller";
+import { AppService } from "./app.service";
+import { SharedModule } from "./shared/shared.module";
+import { HealthModule } from "./health/health.module";
+import { SocketModule } from "./socket/socket.module";
 
-import packageJson from '../package.json';
+import packageJson from "../package.json";
```
small question, was the single quotation to double quotation intentional?

tatarco replied:
it's this missmatch betweeen preetier to eslint, I'll revert

tatarco replied:
its preetier?

djabarovgeorge reviewed (commented):
amazing work!! i left couple of comments for your review.

scopsy committed:
fix: update vercel integration api version

merrcury committed:
feat(root): This introduces telemetry in the OS Build (#6625)

Co-authored-by: Sokratis Vidros <sokratis.vidros@gmail.com>

BiswaViraj committed:
feat(js): update icons and add backdrop-filter (#6629)

SokratisVidros committed:
fix(web): Evaluate IS_V2_ENABLED feature flag correctly.

SokratisVidros committed:
fix(web): Always show the BridgeURL widget

Before this fix, the widget was hidden if the user did not set or cleared the Bridge URL. The same feature flag controlled the top level back button for the workflow page.

tatarco committed:
feat(api): finished PR comments and basic testing

tatarco committed:
feat(api): finished PR comments and basic testing

tatarco committed:
feat(api): finished PR comments and basic testing

scopsy committed:
feat(react-native): Add a react native npm package for hooks (#6556)

ainouzgali committed:
fix(node): add the missing integration override to sms (#6546)

ChmaraX committed:
feat(api): apply member limit per billing plan (#6630)

BiswaViraj committed:
fix(api): response interceptor (#6639)

tatarco committed:
feat(api): fix compilation issue

tatarco committed:
feat(api): fix compilation issue

tatarco committed:
feat(api): fix compilation issue

tatarco committed:
feat(api): fix import

tatarco committed:
feat(api): fix compilation issue
